include/itkAttenuationImageFilter.hxx

```diff
@@ -0,0 +1,225 @@
+{
+  m_RegionSplitter->SetDirection(m_Direction);
+  return m_RegionSplitter.GetPointer();
+}
+
+template <typename TInputImage, typename TOutputImage, typename TMaskImage>
+void
+AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::BeforeThreadedGenerateData()
+{
+  // Verify inputs
+  if (this->GetMaskImage() == nullptr)
+  {
+    itkExceptionMacro("Filter requires a mask image for inclusion estimates!");
+  }
+
+  if (this->GetDirection() > ImageDimension)
+  {
+    itkExceptionMacro("Scan line direction must be a valid image dimension!");
+  }
+
+  if (this->GetSamplingFrequencyMHz() < itk::Math::eps)
+  {
+    itkExceptionMacro("RF sampling frequency was not set!");
+  }
+
+  Superclass::BeforeThreadedGenerateData();
+
+  // Initialize metric image
+  this->GetOutput()->Allocate();
+  this->GetOutput()->FillBuffer(0.0f);
+}
+
+template <typename TInputImage, typename TOutputImage, typename TMaskImage>
+void
+AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::AfterThreadedGenerateData()
+{
+  Superclass::AfterThreadedGenerateData();
+}
+
+template <typename TInputImage, typename TOutputImage, typename TMaskImage>
+void
+AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::DynamicThreadedGenerateData(
+  const OutputRegionType & regionForThread)
+{
+  if (regionForThread.GetNumberOfPixels() == 0)
+  {
+    return;
+  }
+
+  const MaskImageType *  maskImage = this->GetMaskImage();
+  const InputImageType * input = this->GetInput();
+  OutputImageType *      output = this->GetOutput();
+
+  ImageLinearConstIteratorWithIndex<TInputImage> it(input, regionForThread);
+  it.SetDirection(m_Direction);
+  it.GoToBegin();
+
+  unsigned int   inclusionLength;
+  InputIndexType start, end;
+
+  const float scanStepMM = input->GetSpacing()[m_Direction];
+
+  // do the work
+  while (!it.IsAtEnd())
+  {
+    inclusionLength = 0;
+    while (!it.IsAtEndOfLine())
+    {
+      // Advance until an inclusion is found
+      InputIndexType index = it.GetIndex();
+      if (maskImage->GetPixel(index) == m_LabelValue && inclusionLength == 0)
+      {
+        // Step into inclusion
+        start = it.GetIndex();
+      }
+      else if (maskImage->GetPixel(index) != m_LabelValue && inclusionLength > 0)
+      {
+        // Step out of inclusion
+        end = it.GetIndex();
+
+        // Adjust for pixel padding
+        start[m_Direction] += m_PadPixelUpperBounds;
+        end[m_Direction] -= m_PadPixelLowerBounds;
+
+        // Adjust for spatial padding
+        start[m_Direction] += static_cast<unsigned int>(m_PadSpatialUpperBoundsMM / scanStepMM);
+        end[m_Direction] -= static_cast<unsigned int>(m_PadSpatialLowerBoundsMM / scanStepMM);
+
+        // Determine spatial distance between two scanline points and then
+        // adjust for spatial padding
+        unsigned int pixelDistance = end[m_Direction] - start[m_Direction];
+        float        distanceMM = pixelDistance * scanStepMM;
```

Comment:
We don't need either pixelDistance nor distanceMM.

Reply:
Is it guaranteed that `m_FixedEstimationDepth` will always be set? That is, are there valid cases where `m_FixedEstimationDepth` is not set and estimation depths are variable, particularly as the attenuation calculation accommodates variable spatial distances?

Also, it seems like even when `m_FixedEstimationDepth` is set by the user its value will not be exactly the constant spatial distance between pixel centers. We would need to use the closest matching spatial distance between pixels instead, though this could be a one-time calculation.

Reply:
> m_FixedEstimationDepth will always be set

Certainly not. This is one of many options.